rs/types/registry_canister_types/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+use candid::{CandidType, Deserialize};
+use ic_base_types::NodeId;
+use serde::Serialize;
+use std::collections::HashSet;
+use std::fmt;
+use std::net::Ipv4Addr;
+use std::str::FromStr;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum IPv4ConfigError {
+    #[error("Invalid IPv4 address")]
+    InvalidIPv4Address,
+    #[error("Invalid gateway IPv4 address")]
+    InvalidGatewayAddress,
+    #[error("Invalid prefix length")]
+    InvalidPrefixLength,
+    #[error("IP addresses are not in the same subnet")]
+    NotInSameSubnet,
+    #[error("IPv4 address is not a global address")]
+    NotGlobalIPv4Address,
+}
+
+fn is_valid_ipv4_prefix_length(prefix_length: u32) -> bool {
+    prefix_length <= 32
+}
+
+fn is_global_ipv4_address(ipv4_address: &str) -> bool {
+    let ip_address = Ipv4Addr::from_str(ipv4_address).unwrap();
+
+    // IETF RFC 2544
+    let is_benchmarking = u32::from_be_bytes(ip_address.octets()) & (!0 << 17)
+        == u32::from_be_bytes(Ipv4Addr::new(198, 18, 0, 0).octets());
+    // IETF RFC 1112
+    let is_reserved = u32::from_be_bytes(ip_address.octets()) & (!0 << 28)
+        == u32::from_be_bytes(Ipv4Addr::new(240, 0, 0, 0).octets());
+    // IETF RFC 6598
+    let is_shared = u32::from_be_bytes(ip_address.octets()) & (!0 << 22)
+        == u32::from_be_bytes(Ipv4Addr::new(100, 64, 0, 0).octets());
+
+    !is_benchmarking
+        && !ip_address.is_broadcast()
+        && !ip_address.is_documentation()
+        && !ip_address.is_link_local()
+        && !ip_address.is_private()
+        && !ip_address.is_loopback()
+        && !is_reserved
+        && !is_shared
+        && !ip_address.is_unspecified()
+}
+
+// Check that the given IPv4 addresses are all in the same subnet with respect to the subnet mask
```

Comment:
I stared a few seconds at this before I understood that "subnet" means "IPv4 subnet", not "IC subnet". Maybe it should be spelled out as "IPv4 subnet".